### ISSUE TITLE
Failed Invariant 

### ISSUE THREAD

**Issue by winsvega** [ethereum/cpp-ethereum#3571]
The scenario is following.
Node2 is connected as a Peer to Node1. 
Node1 mines a block. Node2 sees it - everything is ok.
Node2 mines a block. Node1 throws failed invariant (in debug)  or not importing a block (in release)

LOG:
```stderr: ☍ ℹ  16:06:09.971|eth     Imported and best 131074  (# 2 ). Has 0 siblings. Route: [#630c055a…, #d908d1f8…]

stderr: ⧫ ℹ  16:06:09.971|eth  1 blocks imported in 0 ms ( 3496.5 blocks/s) in # 2

stderr: ...  16:06:09.971|eth  No transactions to process. 0  pending, 0  queued, 0  future, 0  unverified

stderr: ⧎ ◌  16:06:10.000|p2p|200bb771…|eth/v1.3.0/Linux/g++/Interpreter/Debug/5d022069/  GetBlockHeaders (block (hash): #186846fe… , maxHeaders: 1 , skip: 0 , reverse: 0 )

stderr: ⧎ ◌  16:06:10.000|p2p|4370efc7…|eth/v1.3.0/Linux/g++/Interpreter/Debug/5d022069/  BlocksHeaders ( 1 entries) 

stderr: terminate called after throwing an instance of 'boost::exception_detail::clone_impl<dev::FailedInvariant>'

stderr:   what():  FailedInvariant

stderr: ⧎ ✘  16:06:10.004|p2p|200bb771…|eth/v1.3.0/Linux/g++/Interpreter/Debug/5d022069/  Error reading - Abrupt peer disconnect: End of file
⧎ ◌  16:06:10.005|p2p|200bb771…|eth/v1.3.0/Linux/g++/Interpreter/Debug/5d022069/  Closing peer session :-(

child process exited with code null
```

**Comment by winsvega** [comment]
m_lastImportedBlock is 0 for some reason at instance1
even though it has mined a block.

**Comment by winsvega** [comment]
the issue with faild invariant was happening only on debug build

**Comment by gumb0** [comment]
Sure, checking the invariant is enabled only in debug. But is it the same issue?

**Comment by winsvega** [comment]
I don't really understand what those invariant failures were meant for. 

The issue you reffereing to is about adding a call back for the TestClient class that we developed for solidity testsing and for test rpc

**Comment by gumb0** [comment]
Hm, but we enable this callback in the regular Client, not in the TestClient.
Do you remember what exactly the issue was about?

**Comment by pirapira** [comment]
I'm putting together a fix.  `onBlockImported()` needs to perform a bit more accounting with `m_headers` and `m_bodys`.

**Comment by pirapira** [comment]
After `node main.js` I get
```
[16:22] yh@sri: ~/src/tests/RPCTests(develop|✚1…)$ node main.js     
internal/child_process.js:319
    throw errnoException(err, 'spawn');
    ^
```

Also, I didn't find `ETHPATH` or `TESTPATH` in main.js.

**Comment by pirapira** [comment]
@winsvega will you try branch https://github.com/ethereum/cpp-ethereum/tree/yoichi-3571 ?

**Comment by gumb0** [comment]
^ I would try just doing nothing in the callback in case `!m_headers.empty()`

Looking at the code now I think I see how the problem could happen (without callback) and how callback could fix it and that adding the check for empty `m_headers` shouldn't break it. But yeah, we better should check this RPC test.

What happened:
- initially `m_lastImportedBlock` = 0
- Node1 mines Block1 and sends it to Node2, `m_lastImportedBlock` is still 0 in Node1
- Node1 gets the header of Block2 in `onPeerBlockHeaders`, adds it to `m_headers`
- It's probably empty block, so it adds empty body to `m_bodies` and goes to `collectBlocks`
- `collectBlocks` is supposed to import new block into `BlockQueue`, but it fails to do so because of this check https://github.com/ethereum/cpp-ethereum/blob/develop/libethereum/BlockChainSync.cpp#L582
( `if (headers.first != bodies.first || headers.first != m_lastImportedBlock + 1) return;`) 
(the check makes sense, we don't start to import sequence of blocks until we download the sequence "connected" to last imported block)
- then we get the failure of this invariant: 
```
if (isSyncing() && !m_headers.empty() &&  m_lastImportedBlock >= m_headers.begin()->first)
		BOOST_THROW_EXCEPTION(FailedInvariant() << errinfo_comment("Header is too old"));
```

What could be better fix - I don't know yet

**Comment by gumb0** [comment]
Ah no wait, that invariant should be fine

**Comment by winsvega** [comment]
RPC tests main.js is not supported. although it should return no errors. It fails on azure to mine a block for unknown reason. RPCTests was designed to check cpp client rpc methods and functionality. 
like client syncing. 

the basic tests are starting 2 nodes. mine a block. send a transaction. mine a block with that transaction and check that block is propagated to the second instance of a client. it works locally.

**Comment by pirapira** [comment]
Now I don't see an Invariant Failure on this branch: https://github.com/ethereum/cpp-ethereum/tree/yoichi-3571

**Comment by gumb0** [comment]
Ok, I've got the RPC tests running and now `TEST_mineBlockOnNode1` & `TEST_mineBlockOnNode2` are still failing on latest develop with error "You need to set an author in order to seal!" (RelWithDebInfo build)

**Comment by winsvega** [comment]
that is a strange issue. it was happening before. 
https://github.com/ethereum/cpp-ethereum/blob/develop/libethereum/ClientTest.cpp#L64

**Comment by gumb0** [comment]
After manually setting author with `eth -a`, `TEST_mineBlockOnNode1` & `TEST_mineBlockOnNode2` pas successully, both with `onBlockImported` callback disabled and enabled.
(then it fails on some other test with not related error)

So I don't see FailedInvariant with RPCTests at all.
And probably callback's purpose was to fix another issue.

**Comment by gumb0** [comment]
Ah no, `TEST_getBlockHashOnNode1` is passing with enabled callback and is failing with FailedInvariant in debug with disabled callback, I see now.

**Comment by gumb0** [comment]
So yes, without callback this invariant check in `BlockChainSync` fails at the moment when the header is received in the Node1 from Node2 (before the code of `onPeerBlockHeaders`):
```
	if (isSyncing() && m_host.chain().number() > 0 && m_haveCommonHeader && m_lastImportedBlock == 0)
		BOOST_THROW_EXCEPTION(FailedInvariant() << errinfo_comment("Common block not found"));
```

All variable values make sense at this moment, the condition doesn't really make sense, I'm not sure yet how to fix it without the callback. The easiest would be to remove the condition of course.
